benchmark/submit.py

```
@@ -0,0 +1,117 @@
+        d = json_content(response)
+        d.addCallback(lambda content: (response, content))
+        return d
+
+    req.addCallback(get_response_content)
+
+    def process_response(response_and_content):
+        (response, content) = response_and_content
+        if response.code != CREATED:
+            raise SubmitFailure(response.code, response.phrase,
+                                content['message'])
+        else:
+            return content['id']
+
+    req.addCallback(process_response)
+    return req
+
+
+class Options(usage.Options):
+    longdesc = ("Submit a benchmark result provided as a JSON document "
+                "on the standard input to the benchmark server.")
+
+    optParameters = [
+        ['address', None, None, "The address of the benchmark server."],
+        ['port', None, 8888, "The port of the benchmark server.", int],
+    ]
+
+    def postOptions(self):
+        if self.get('address') is None:
+            raise usage.UsageError("The address must be specified.")
+
+
+def main(reactor, args):
+    try:
+        options = Options()
+        options.parseOptions(args)
+    except usage.UsageError as e:
+        print(e.args[0], file=sys.stderr)
+        print('', file=sys.stderr)
+        print(str(options), file=sys.stderr)
+        raise SystemExit(1)
+
+    server_url = b"http://{}:{}".format(
+        options['address'], options['port'],
+    )
+
+    try:
+        result_input = sys.stdin.read()
+        result = json.loads(result_input)
+    except ValueError as e:
+        print(result_input, file=sys.stderr)
+        print(
+            "Standard input is not valid JSON document: {}".format(e.message),
+            file=sys.stderr,
+        )
+        raise SystemExit(1)
+
+    d = submit(server_url, result)
+
+    def succeeded(id):
+        print("Assigned result ID: {}".format(id))
+
+    def failed(failure):
+        print("Failed to submit the result: {}".format(failure.value),
+              file=sys.stderr)
```

Interesting tests! I decided to do the nice reporting only for errors returned by the REST API, that is, for `SubmitFailure` and used `Failure.trap()` for that. I treat all other exceptions as unexpected conditions and let them propagate all the way up, they are reported in a more legible fashion now. I guess tat I could have the nice reporting for Twisted network errors as well, but it seems that there are quite a few of them and they do not seem to have a distinct common base.